File: base/Sources/logic_node.ts

type logic_node_t = {
	inputs?: logic_node_input_t[];
	outputs?: logic_node_t[][];
	get?: (self: any, from: i32, done: (a: any)=>void)=>void;
	get_as_image?: (self: any, from: i32, done: (img: image_t)=>void)=>void;
	get_cached_image?: (self: any)=>image_t ;
	set?: (self: any, value: any)=>void;
};

type logic_node_input_t = {
	node?: any; // logic_node_t
	from?: i32; // Socket index
};

function logic_node_create(): logic_node_t {
	let n: logic_node_t = {};
	n.inputs = [];
	n.outputs = [];
	n.get = logic_node_get;
	n.get_as_image = logic_node_get_as_image;
	n.get_cached_image = logic_node_get_cached_image;
	n.set = logic_node_set;
	return n;
}

function logic_node_add_input(self: logic_node_t, node: logic_node_t, from: i32) {
	self.inputs.push(logic_node_input_create(node, from));
}

function logic_node_add_outputs(self: logic_node_t, nodes: logic_node_t[]) {
	self.outputs.push(nodes);
}

function logic_node_get(self: logic_node_t, from: i32, done: (a: any)=>void) {
	done(null);
}

function logic_node_get_as_image(self: logic_node_t, from: i32, done: (img: image_t)=>void) {
	done(null);
}

function logic_node_get_cached_image(self: logic_node_t): image_t {
	return null;
}

function logic_node_set(self: logic_node_t, value: any) {}

function logic_node_input_create(node: logic_node_t, from: i32): logic_node_input_t {
	let inp: logic_node_input_t = {};
	inp.node = node;
	inp.from = from;
	return inp;
}

function logic_node_input_get(self: logic_node_input_t, done: (a: any)=>void) {
	self.node.base.get(self.node, self.from, done);
}

function logic_node_input_get_as_image(self: logic_node_input_t, done: (img: image_t)=>void) {
	self.node.base.get_as_image(self.node, self.from, done);
}

function logic_node_input_set(self: logic_node_input_t, value: any) {
	self.node.base.set(self.node, value);
}

File: paint/sources/logic_node.c

#include "global.h"

logic_node_t *logic_node_create(logic_node_ext_t *ext) {
	logic_node_t *n = GC_ALLOC_INIT(logic_node_t, {0});
	n->inputs       = any_array_create_from_raw((void *[]){}, 0);
	n->outputs      = any_array_create_from_raw((void *[]){}, 0);
	n->ext          = ext;
	return n;
}

logic_node_input_t *logic_node_input_create(logic_node_ext_t *node, i32 from) {
	logic_node_input_t *inp = GC_ALLOC_INIT(logic_node_input_t, {0});
	inp->node               = node;
	inp->from               = from;
	return inp;
}

void logic_node_add_input(logic_node_t *self, logic_node_ext_t *node, i32 from) {
	any_array_push(self->inputs, logic_node_input_create(node, from));
}

void logic_node_add_outputs(logic_node_t *self, logic_node_t_array_t *nodes) {
	any_array_push(self->outputs, nodes);
}

void * logic_node_get(logic_node_t *self, i32 from) {
	if (self->get != NULL) {
		return self->get(self->ext, from);
	}
	return NULL;
}

gpu_texture_t *logic_node_get_as_image(logic_node_t *self, i32 from) {
	if (self->get_as_image != NULL) {
		return self->get_as_image(self->ext, from);
	}
	return NULL;
}

void logic_node_set(logic_node_t *self, void * value) {
	if (self->set != NULL) {
		self->set(self->ext, value);
	}
}

logic_node_value_t *logic_node_input_get(logic_node_input_t *self) {
	return logic_node_get(self->node->base, self->from);
}

gpu_texture_t *logic_node_input_get_as_image(logic_node_input_t *self) {
	return logic_node_get_as_image(self->node->base, self->from);
}

void logic_node_input_set(logic_node_input_t *self, f32_array_t *value) {
	logic_node_set(self->node->base, value);
}
